### ISSUE TITLE
Erroneous response on getDescGeneric() may lead to improper memory access 

### ISSUE THREAD

**Issue by samuelfmlourenco** [bloguetronica/mcp2210#11]
This is similar to the case reported [here](https://github.com/bloguetronica/mcp2210-qt/issues/5). However, unlike Qt, C++ by itself might not have the mechanisms to detect this, and the program may even run fine, depending on the compiler that is being used. Still, improper memory access, outside the bounds of the vector, should be dealt with, even if it doesn't lead to a crash. Here is the function as implemented:

```
// Private generic function that is used to get any descriptor
std::u16string MCP2210::getDescGeneric(uint8_t subcomid, int &errcnt, std::string &errstr)
{
    std::vector<uint8_t> command{
        GET_NVRAM_SETTINGS, subcomid  // Header
    };
    std::vector<uint8_t> response = hidTransfer(command, errcnt, errstr);
    size_t maxLength = 2 * DESC_MAXLEN + 2;  // Maximum descriptor length in bytes
    size_t length = (response[4] > maxLength ? maxLength : response[4]) - 2;  // Descriptor internal length
    std::u16string descriptor;
    for (size_t i = 0; i < length; i += 2) {
        descriptor += static_cast<char16_t>(response[i + 7] << 8 | response[i + 6]);  // UTF-16LE conversion as per the USB 2.0 specification
    }
    return descriptor;
}
```

The fix, again, is very simple. By just adding an extra condition to the for loop, it is then possible to prevent the error described above:

```
// Private generic function that is used to get any descriptor
std::u16string MCP2210::getDescGeneric(uint8_t subcomid, int &errcnt, std::string &errstr)
{
    std::vector<uint8_t> command{
        GET_NVRAM_SETTINGS, subcomid  // Header
    };
    std::vector<uint8_t> response = hidTransfer(command, errcnt, errstr);
    size_t maxLength = 2 * DESC_MAXLEN + 2;  // Maximum descriptor length in bytes
    size_t length = (response[4] > maxLength ? maxLength : response[4]) - 2;  // Descriptor internal length
    std::u16string descriptor;
    for (size_t i = 0; i < length && i < COMMAND_SIZE - 6; i += 2) {  // Extra condition added here!
        descriptor += static_cast<char16_t>(response[i + 7] << 8 | response[i + 6]);  // UTF-16LE conversion as per the USB 2.0 specification
    }
    return descriptor;
}
```

**Comment by samuelfmlourenco** [comment]
Confirmed segmentation fault when reading the descriptor. This bug is severe.

**Comment by samuelfmlourenco** [comment]
Solution as implemented in version 1.2.0:

```
// Private generic function that is used to get any descriptor
std::u16string MCP2210::getDescGeneric(uint8_t subcomid, int &errcnt, std::string &errstr)
{
    std::vector<uint8_t> command{
        GET_NVRAM_SETTINGS, subcomid  // Header
    };
    std::vector<uint8_t> response = hidTransfer(command, errcnt, errstr);
    size_t maxSize = 2 * DESC_MAXLEN;  // Maximum size of the descriptor in bytes (the zero padding at the end takes two more bytes)
    size_t size = response[4] - 2;  // Descriptor actual size, excluding the padding
    size = size > maxSize ? maxSize : size;  // This also fixes an erroneous result due to a possible unsigned integer rollover (bug fixed in version 1.2.0)
    std::u16string descriptor;
    for (size_t i = 0; i < size; i += 2) {
        descriptor += static_cast<char16_t>(response[i + PREAMBLE_SIZE + 3] << 8 | response[i + PREAMBLE_SIZE + 2]);  // UTF-16LE conversion as per the USB 2.0 specification
    }
    return descriptor;
}
```
